Feature/s3 custom endpoints

Konstantinos-10 opened the pull request:
## Description

This PR adds support for custom S3 endpoints in the MindsDB S3 handler, enabling users to connect to S3-compatible services like MinIO. The changes include:

- Added `endpoint_url` parameter to the S3 handler configuration
- Implemented proper endpoint handling in the S3 client initialization
- Added comprehensive test suite for custom endpoint functionality
- Improved error handling and connection management
- Added detailed documentation for custom endpoint usage

This enhancement allows users to connect to any S3-compatible service by specifying their custom endpoint URL, making the S3 handler more flexible and versatile.

## Type of change

- [x] ⚡ New feature (non-breaking change which adds functionality)

## Verification Process

To ensure the changes are working as expected:

- [x] Test Location: `mindsdb/integrations/handlers/s3_handler/tests/test_s3_endpoints.py`
- [x] Verification Steps:
  1. Set up a MinIO instance running on localhost:9000
  2. Create a test bucket named 'test-bucket'
  3. Configure MinIO with access key 'minioadmin' and secret key 'minioadmin'
  4. Run the test suite which will:
     - Test connection to MinIO
     - List available buckets
     - Read from a test file
     - Write a new file
     - Clean up test files
  5. Verify that all operations complete successfully with the custom endpoint

## Checklist:

- [x] My code follows the style guidelines(PEP 8) of MindsDB.
  - All code changes follow PEP 8 standards
  - Proper indentation and spacing maintained
  - Consistent naming conventions used

- [x] I have appropriately commented on my code, especially in complex areas.
  - Added comments explaining endpoint handling logic
  - Documented test cases and their purposes
  - Added inline comments for complex operations

- [x] Necessary documentation updates are either made or tracked in issues.
  - Updated S3 handler documentation to include endpoint_url parameter
  - Added examples for custom endpoint usage
  - Documented MinIO configuration requirements

- [x] Relevant unit and integration tests are updated or added.
  - Created new test file `test_s3_endpoints.py`
  - Added comprehensive test cases for custom endpoints
  - Included timeout handling and error cases
  - Added cleanup procedures in tests

I have read the CLA Document and I hereby sign the CLA

ZoranPandovski commented:
recheckcla

ZoranPandovski commented on mindsdb/integrations/handlers/s3_handler/tests/setup_test_env.sh:
```diff
@@ -0,0 +1,31 @@
+#!/bin/bash
```
We don't need this sh script, for now you can remove it

ZoranPandovski commented on mindsdb/integrations/handlers/s3_handler/README.md:
```diff
@@ -100,4 +100,179 @@ This table will return all objects regardless of the file format, however, only
         * Correct: SELECT * FROM integration.\`travel/travel_data.csv\`
-</Warning>
\ No newline at end of file
+</Warning>
```
Let's update this doc for now you can add an example with the custom config. If we keep it like this it looks like we duplicate the same page

ZoranPandovski commented on mindsdb/integrations/handlers/s3_handler/s3_handler.py:
```diff
@@ -275,20 +293,26 @@ def add_data_to_table(self, key, df) -> None:
-
-            # upload
+            # Create a temporary table with the data
+            connection.execute("CREATE TABLE tmp_table AS SELECT * FROM df")
```
How this will work? What is the df?

Konstantinos-10 replied:
The variable df is the Pandas DataFrame passed into the add_data_to_table() method. DuckDB allows querying directly from in-memory Pandas DataFrames using SQL. The line CREATE TABLE tmp_table AS SELECT * FROM df uses this feature to create a temporary DuckDB table from the DataFrame before copying it to the target S3 path. It works because DuckDB recognizes Python DataFrames by name in the local context.

ZoranPandovski commented on mindsdb/integrations/handlers/s3_handler/s3_handler.py:
```diff
@@ -163,6 +163,20 @@ def _connect_duckdb(self, bucket):
+        if 'endpoint_url' in self.connection_data:
+            endpoint = self.connection_data['endpoint_url']
+            # Remove http:// or https:// prefix if present
+            if endpoint.startswith('http://'):
```
Maybe we improve this using urllib.parse

Konstantinos-10 replied:
I used urllib.parse and changed the code like this:

```        # Set custom endpoint if provided
        if 'endpoint_url' in self.connection_data:
            # Extract host and port from endpoint URL
            raw_url = self.connection_data['endpoint_url']
            parsed_url = urlparse(raw_url)
            endpoint_host = (parsed_url.netloc or parsed_url.path).rstrip('/')  # handles cases like "http://localhost:9000" or "localhost:9000"

            duckdb_conn.execute(f"SET s3_endpoint='{endpoint_host}'")
            duckdb_conn.execute("SET s3_url_style='path'")
            duckdb_conn.execute("SET s3_use_ssl='false'")
```

ZoranPandovski reviewed (changes requested):
Thanks, please check the comments

Konstantinos-10 commented:
I've updated the handler to use urllib.parse, and I’ve added a test to validate the parsing logic. Let me know if there’s anything else you’d like adjusted!

martyna-mindsdb commented:
@Konstantinos-10 
Are you still working on this? If so, please resolve the conflicts in this PR.

Konstantinos-10 commented:
yes of course i just had my exams @martyna-mindsdb
